Sharing dashboard subscriptions slack performance

escherize opened the pull request:
Fixes backend piece for: #18774

This PR adds caching to the Backend for slack data that may take a long time to query for. It uses the settings capability of the app for a few reasons, mainly that the cached data can be shared in a multi-instance setup of Metabase. Slack only allows a linear scan through their users and conversations apis, so we need to walk all pages before sending the frontend the data.

#### When we receive a GET on `form_input` (always fast, potentially stale)
```mermaid
sequenceDiagram
    autonumber
    title: [Fetch Slack user+channels] finishes before [[slack/timeout-ms]]
    participant F as Frontend
    participant B as Backend
    participant S as Slack API
    activate B
    F->>B: request results  
    Note right of B: new request for this endpoint, re-triggers Fetch and Cache slack user+channels
    B->>F: return cached results
    B->>B: request kicks off Fetch and Cache slack user+channels
   
    loop Fetch Slack user+channels
        B->>S: request_results()
        alt
        S->>B: results + cursor
        B->>S: request_results(cursor)
    end
    S->>B: results (no cursor)
    deactivate B
    B->>B: refresh cache
    activate B
    end
    deactivate B
```

We get an [almost 2x speedup](https://github.com/metabase/metabase/pull/20570/commits/b6cacef6315fbd69d1d24cbcb595f79183193559) from parallelizing the requests to slack's user and conversation endpoints for `form_input_refresh`

### When we receive a GET on `form_input_refresh` (potentially slow, always up to date)
```mermaid
sequenceDiagram
    autonumber
    title: [Fetch Slack user+channels] finishes before [[slack/timeout-ms]]
    participant F as Frontend
    participant B as Backend
    participant S as Slack API
    activate B
    F->>B: request results  
    Note right of B: new request for this endpoint, re-triggers Fetch and Cache slack user+channels
   
    B->>B: request kicks off Fetch and Cache slack user+channels
   
    loop Fetch Slack user+channels
        B->>S: request_results()
        alt
        S->>B: results + cursor
        B->>S: request_results(cursor)
    end
    S->>B: results (no cursor)
    deactivate B
    B->>B: refresh cache
    activate B
    end
    B->>F: return cached results
    deactivate B
```

escherize committed:
Get slack users and conversations simulatenously

- instead of waiting for the channels, and then getting the users

escherize committed:
Add cache+timeout capability to conversations-list

escherize committed:
extend async approach to users-list

- return timeout+cache warning
- adds tests

escherize commented on src/metabase/api/pulse.clj:
```diff
@@ -135,11 +136,15 @@
-                                          (str \@ (:name user)))]
-                     (assoc-in chan-types [:slack :fields 0 :options] (concat slack-channels slack-users)))
+                   (let [[conversations users] (map deref [(future (slack/conversations-list-timeout))
+                                                           (future (slack/users-list-timeout))])
```
this way we don't need to wait for the conversations list to finish to start getting the users list.

These *-timeout functions return a map like `{:result [...] :timeout true}` if the timeout was hit and we need to return the cached version.

escherize committed:
updating docstrings + adding more tests

escherize committed:
Add cache clear and seed slack token settings on-change

- add dox for :on-change for defsettings
- fixing tests

escherize committed:
Fixing tests

- cannot pass a declared (not def'd) value
- some cleanup

escherize committed:
remove unused function

escherize commented on src/metabase/api/slack.clj:
```diff
@@ -22,10 +22,9 @@
                         {:errors {:slack-app-token (tru "invalid token")}}))))
     (slack/slack-app-token slack-app-token)
     (when slack-app-token
```
removed a do

escherize committed:
remove need for slack-configured from timeout fns

- update some docstrings

escherize committed:
Merge branch 'master' into sharing-dashboard-subscriptions-slack-performance

escherize committed:
pre-populate slack caches

- on app init

escherize committed:
introduce blocking (and cache resetting) endpoint

- this one does not timeout, meant to be called when a user hits a
  button like: "Refetch my slack channels"

escherize committed:
removes -timeout versions of users/channels fn

- the endpoint now returns the cached slack data
- removes tests

escherize committed:
fixes routing layer

escherize committed:
remove trailing whitespace

escherize committed:
add missing docstring

escherize committed:
move the cached slack data into a setting

escherize committed:
add missing docstring

escherize committed:
fix typo

escherize committed:
update form-input-test

escherize committed:
refresh cache in test

escherize commented:
In a multi-instance horizontally scaled metabase setup, a client is not guaranteed to be sending api requests to the same server. So I stored the slack user and conversation caches as Settings. There are some benefits to storing them as settings. If they're persisted there's no need to re-seed when the app restarts. Also we won't have to worry about them "disappearing" when a client is directed to different servers.

I’ve also been [discussing UX details with maz and bruno](https://metaboat.slack.com/archives/C01NT034X2B/p1645552300192499). They'd like to see a way for users to "Get my updated slack channels" which would get the new data. So I will add a route `/form_input_refresh` which blocks until the slack data has been retrieved, refreshes the cache, and returns. I am working to get the frontend piece specified and handed over now.

## What happens now:

After clicking Share on a dashboard, The slack endpoint can take a very long time to load, then there is a problem wherein the user cannot select email or slack. 

## What should be done?

- [x] `/form_input` returns cached version in the same shape
- [x] `/form_input_refresh` behaves like the old endpoint (which can take a long time, but garuntees the complete list)
- [x] reseed cache on app startup
- [x] reseed cache asynchronistically whenever `/form_input` is hit

## Things to do on the frontend

- [ ] Render the "Create a dashboard subscription" sidepanel when the `/form_input`request is ongoing.
- [ ] After clicking the "Send it to Slack" card, A spinner with message "Refreshing the list of channels…" to indicate an inflight response to `/form_input` or `/form_input_refresh`.
- [ ] On the list view, add some text like “Not seeing the channel you’re looking for? Refresh this list.” with a button that sends a get to `/form_input_refresh`. This endpoint behaves just like `/form_input`, used to, where it blocks and returns the user's gaurunteed-up-to-date slack channels. (Because it was taking so long, `/form_input` has been changed to short-circuit and return cached data after 0.5 seconds.)

escherize commented on src/metabase/core.clj:
```diff
@@ -122,6 +123,9 @@
 
+  (when (slack/slack-configured?)
+    (slack/refresh-cache!))
+
```
I wasn't sure if this was appropriate, all it does is (potentially) fire off some http requests, and set the slack-cache setting.

noahmoss replied:
Won't this block until the Slack requests return? If so we probably shouldn't do it during startup.

escherize replied:
Good catch, I'll wrap this in a future, I didn't want to block here either

escherize committed:
remove un-needed anonymous fn wrapping

escherize committed:
refresh slack cache when user sets new valid token

escherize committed:
refresh the cache when reading from it

- the more its used the more up to date it will be

noahmoss commented on src/metabase/integrations/slack.clj:
```diff
@@ -36,14 +36,17 @@
     (if (str/starts-with? channel-name "#") (subs channel-name 1) channel-name)))
 
+(defsetting slack-cache
+  "A cache shared between instances for storing a user's slack channels and users."
```
nit: an _instance's_ channels & users?

noahmoss commented on src/metabase/integrations/slack.clj:
```diff
@@ -165,24 +168,36 @@
+       vec))
+
+(defn refresh-cache!
+  "Refreshes users and conversations in slack-cache. If called with no args, finds both in paralell, and sets the cache."
```
nit: _parallel_ spelling

noahmoss reviewed (changes requested):
Code lgtm, and storing the cache in an internal setting seems reasonable, but I'm curious what @camsaul thinks. Particularly if there could be any issues for Slack workspaces with huge (1M+) channel lists.

I think the cache should be cleared if the Slack connection is disabled. Don't need to store that data longer than we have to. Requesting changes for this reason.

Also, is there any way to directly test the caching behavior?

escherize committed:
code review responses + removes 2-arity refresh-cache!

- clear out cache when slack connection is broken or unset
- add to docstring for slack settings route
- wrap init refresh-cache! with a future
- re-spell parallel

escherize committed:
do not block when returning cached slack data

camsaul commented on src/metabase/api/pulse.clj:
```diff
@@ -117,6 +116,35 @@
+                 ;; quite a while).
+                 :else
+                 (try
+                   (let [[conversations users] (map deref [(future (slack/conversations-list))
+                                                           (future (slack/users-list))])
```
`pmap` does the exact same thing as this

camsaul commented on src/metabase/core.clj:
```diff
@@ -122,6 +123,9 @@
       ;; otherwise update if appropriate
       (sample-data/update-sample-database-if-needed!)))
 
+  (when (slack/slack-configured?)
+    (future (slack/refresh-cache!)))
```
If I spin up 4 instances is this going to trigger four simultaneous refreshes? Would it be better to schedule a Quartz job to fire immediately so it only runs on one instance?

camsaul replied:
As mentioned previously, I'm not sure this specific code here makes sense -- it doesn't do anything on first launch _before_ Slack is configured, and it doesn't do much besides that other than triggering one-off updates when an instance is restarted. If an instance runs for a few months, this isn't enough on its own to keep things up to date on its own

escherize replied:
The specific problem I am trying to solve with this piece is for users who have slack configured today. When they upgrade their slack users/convo cache will be empty and that will cause the `/form_input` route to act strangely. 

Granted the endpoint will fix itself once the route is hit + (time it takes to crawl their slack data), but until then they'll get no results. Would like to hear if you have another fix for that case.

I considered using the Quartz scheduler, I think you're right that that's a better way to handle this.

camsaul commented on src/metabase/api/pulse.clj:
```diff
@@ -117,6 +116,35 @@
+                 (try
+                   (let [[conversations users] (map deref [(future (slack/conversations-list))
+                                                           (future (slack/users-list))])
+                         _                     (slack/slack-cache {:conversations conversations :users users})
```
If we only need the `:name` of the channels and users we should only cache those rather than the entire gigantic response that we get back from the Slack API

camsaul commented on src/metabase/api/pulse.clj:
```diff
@@ -117,6 +116,35 @@
+                 ;; if we have Slack enabled, block until we gather the entire list of conversations + users (can be
+                 ;; quite a while).
+                 :else
+                 (try
+                   (let [[conversations users] (map deref [(future (slack/conversations-list))
+                                                           (future (slack/users-list))])
+                         _                     (slack/slack-cache {:conversations conversations :users users})
+                         slack-channels        (for [channel conversations] (str \# (:name channel)))
+                         slack-users           (for [user users] (str \@ (:name user)))]
+                     (assoc-in chan-types [:slack :fields 0 :options] (concat slack-channels slack-users)))
+                   (catch Throwable e
+                     (assoc-in chan-types [:slack :error] (.getMessage e)))))}))
```
Some thoughts:
1. Since `conversations-list` and `users-list` both return lazy sequences you're not actually running these on separate threads the way you think. You're fetching the first page for each on separate threads, but these aren't actually ultimately realized until you call `slack/slack-cache` below, at which point the lazy sequences get realized on a single thread.
2. Since we only ever use the `@name` or `#name` info we should calculate that here and store _that_ in the cache rather than storing the entire enormous Slack API response in the cache and then calculating the same stuff when we pull the cached results out. 
3. You should be converting channels/users to `#name` or `@name` immediately as the lazy seq is realized rather than realizing the whole thing and _then_ converting them -- as @noahmoss mentioned for a huge Slack instance with millions of Users that's going to make a huge difference in terms of the peak amount of memory this uses (1 million giant maps is a lot bigger than 1 million strings)

Here's one way you might consider reworking it.

The basic idea is to fully realize `(slack/conversations)` and `(slack/users-list)` on separate threads. We transduce those and convert them to `#name` or `@name` immediately on the respective separate threads, then concat those together with `mapcat` (this itself is a lazy sequence but that ultimately won't make a difference here since we're fully realizing stuff in the background threads). 

You can store this concatenated list in the cache rather than the entire API response. 

Let me know if this makes sense.

```suggestion
                 (try
                   (let [channels (mapcat deref [(future (mapv #(str \# (:name %))) (slack/conversations-list))
                                                 (future (mapv #(str \@ (:name %))) (slack/users-list))])]
                     ;; update the cached channels
                     (slack/slack-cache channels)
                     (assoc-in chan-types [:slack :fields 0 :options] channels))
                   (catch Throwable e
                     (assoc-in chan-types [:slack :error] (.getMessage e)))))}))
```

camsaul replied:
I guess my comments about this not really being truly multithreaded aren't 100% true since you changed the underlying functions to fully realized everything (by wrapping in `(vec ...)`). But I think that's overall a bad change since we'd be forced to realize the entire API response and couldn't discard unneeded bits as we go. I'll add some more comments about why to that part of the diff

camsaul commented on src/metabase/integrations/slack.clj:
```diff
@@ -36,14 +36,17 @@
   (when-not (str/blank? channel-name)
     (if (str/starts-with? channel-name "#") (subs channel-name 1) channel-name)))
 
+(defsetting slack-cache
```
I would suggest naming this something like `cached-channel-and-user-names` for a few reasons:

1. `slack-` doesn't really give us any extra information, we know this has to do with Slack because it's in a `.slack` namespace -- see https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#dont-repeat-namespace-alias-in-function-names for more info.
2. `cache` is way to vague. `cache` _of what?_

escherize replied:
I think you're exactly right on 2, I should have renamed that.

Also, I need some clarification on 1. Because this value is a `defsetting`, it needs a globally unique name. And looking at other defsettings the linked section of the guide isn't being followed. Here are a few examples I found:

- [snowplow.clj#L30](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/analytics/snowplow.clj#L30)
- [snowplow.clj#L38](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/analytics/snowplow.clj#L38)
- [geojson.clj#L83](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/api/geojson.clj#L83)
- [geojson.clj#L83](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/api/geojson.clj#L83)
- [geojson.clj#L83](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/api/geojson.clj#L83)
- [ldap.clj#L16](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L16)
- [ldap.clj#L21](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L21)
- [ldap.clj#L24](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L24)
- [ldap.clj#L29](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L29)
- [ldap.clj#L38](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L38)
- [ldap.clj#L41](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/ldap.clj#L41)
- [slack.clj#L16](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/slack.clj#L16)
- [slack.clj#L22](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/slack.clj#L22)
- [slack.clj#L27](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/slack.clj#L27)

At a glance, the majority of the `defsettings` have their namespace prefixed into them. I figured the reason for this is that when reading _the code_ it's clear what namespace the var came from, but when looking at settings e.g. in the database I don't think you get the same affordance. Is that correct? I could use some more context. Maybe `slack-cached-channel-and-user-names` is the right name.

camsaul commented on src/metabase/integrations/slack.clj:
```diff
@@ -36,14 +36,17 @@
   :setter (fn [channel-name]
             (setting/set-value-of-type! :string :slack-files-channel (process-files-channel-name channel-name))))
 
-(def ^:private ^String slack-api-base-url "https://slack.com/api")
-
```
What's the reason for removing the constant and replacing it with magic strings below?

escherize replied:
This was only used in one place, so the abstraction about what `slack-api-base-url` is doesn't add much value. Now, readers of `do-slack-request` don't need to pop up and check what `slack-api-base-url`'s value is.

escherize replied:
Before, it was marked private since it is un-needed in other namespaces. Now its level of privacy is higher, since it's un-usable from anywhere besides `do-slack-request`.

camsaul commented on src/metabase/integrations/slack.clj:
```diff
@@ -154,7 +157,7 @@
   [& {:as query-parameters}]
   (let [params (merge {:exclude_archived true, :types "public_channel"} query-parameters)]
-    (paged-list-request "conversations.list" :channels params)))
+    (vec (paged-list-request "conversations.list" :channels params))))
```
As mentioned above, there are a few reasons why I think forcing this to be fully realized _here_ is a bad idea.

1. This forces us to have the entire API response in memory, even if we only need one key from each object in the results (as is the actual case). This means this ends up taking a huge amount of memory for big API responses
2. If we only wanted one or two items this still ends up fetching potentially millions of items. `(take 1 infinite-lazy-sequence)` will just return the first item and will not force the rest of the `infinite-lazy-sequence` to be realized. `(take 1 (conversations-list))` will never fetch more than one page if `(conversations-list)` is lazy; if it's not we're always forced to fetch every single page regardless of how many we actually _want_.

camsaul commented:
Have you considered having the `form_input` endpoint automatically trigger a cache update when needed, instead of asking the frontend to do it manually? You could store a separate internal Setting called `channel-and-username-list-last-updated` or something like that, and then `form_input` could return cached results if the cache is not too old, otherwise block and update cached results first, then return those. Then you don't need `form_input_refresh` at all unless you wanted to have some way for people to manually trigger a refresh

camsaul commented on src/metabase/api/pulse.clj:
```diff
@@ -117,6 +116,35 @@
   ;; return updated Pulse
   (pulse/retrieve-pulse id))
 
+(api/defendpoint GET "/form_input_refresh"
+  "Provides relevant configuration information and user choices for creating/updating Pulses, can block for quite a long
+  amount of time if the conversations-list or the users-list is huge."
+  []
+  (let [chan-types (-> channel-types
```
It seems like a malicious user could hit this endpoint a bunch of times over and over and nuke an instance -- it seems like we should probably limit this somehow. If someone hit refresh 10 seconds ago there is nothing another refresh is likely to do anyway, right?

camsaul reviewed (changes requested):
I left a bunch of notes on how this can be reworked to simplify the code a bit and dramatically reduce the memory usage as well as the size of the cache in the application DB, and a handful of style notes. I also left some notes on how you can have this work without expecting the FE client to manually trigger cache refreshes (which still need to be done if I understand correctly). Of course, that would mean requests that need to fetch new results would end up being pretty slow, so if we wanted to pre-warm the cache it would probably best be done with scheduled Quartz jobs.

Ideally I think we wouldn't want to have a `form_input_refresh` endpoint at all since it could easily be abused, but if we do have one we definitely need to add some code to limit it so a crazy user doesn't go in and try to trigger a thousand simultaneous cache refreshes and nuke the entire MB instance. 

It might actually be best to have Slack result refreshes _always_ be done as Quartz jobs, so only one instance at a time will do it (which will eliminate duplicate work) and have the `form_input_refresh` endpoint just schedule a new job if one is not already in progress.

Let me know if those suggestions make sense.

RE @noahmoss's comments:

> Storing the cache in an internal setting seems reasonable, but I'm curious what @camsaul thinks. Particularly if there could be any issues for Slack workspaces with huge (1M+) channel lists.

This is a good point... it's probably a good idea to have some sort of upper limit just in case someone has a gigantic instance. Of course, we'd still run into problems if someone wanted to send a message to `@xena-the-warrior-princess` if she doesn't show up in the list, but I think that should be solved in a different way anyway (perhaps by letting people type arbitrary Slack channel/user names in addition to choosing them from a drop-down.)

> I think the cache should be cleared if the Slack connection is disabled. Don't need to store that data longer than we have to. Requesting changes for this reason.

👍 on this, good catch @noahmoss. Someone can potentially change Slack connections as well -- we don't want to show results from the old connection if that happens.

camsaul commented:
One more thought:

1. I launch a brand-new Metabase instance.
2. Slack users/channels cache will be empty. Code to populate it on launch can't run with no credentials.
3. Configure Slack
4. Slack users/channels cache is still empty, right?
5. I hit the `form_input` endpoint, I get blank results

I think that's a good case for making cache refresh happen automatically if needed when you hit the `form_input` endpoint rather than expecting people to manually hit a refresh button somewhere. And it would eliminate the need to have that code to update the cache inside `metabase.core`

escherize committed:
Merge branch 'master' into sharing-dashboard-subscriptions-slack-performance

escherize committed:
rename cache defsetting

escherize commented:
> One more thought:
> 
>     1. I launch a brand-new Metabase instance.
> 
>     2. Slack users/channels cache will be empty. Code to populate it on launch can't run with no credentials.
> 
>     3. Configure Slack
> 
>     4. Slack users/channels cache is still empty, right?
> 
>     5. I hit the `form_input` endpoint, I get blank results
> 
> 
> I think that's a good case for making cache refresh happen automatically if needed when you hit the `form_input` endpoint rather than expecting people to manually hit a refresh button somewhere. And it would eliminate the need to have that code to update the cache inside `metabase.core`

The cache repopulates whenever a new slack token is deemed valid [here](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/api/slack.clj#L36).

escherize commented:
> Storing the cache in an internal setting seems reasonable, but I'm curious what @camsaul thinks. Particularly if there could be any issues for Slack workspaces with huge (1M+) channel lists.

This point stands, but one subtle misstatement is that [if slack accounts have more than 10k users or channels, we skip the rest.](https://github.com/metabase/metabase/blob/sharing-dashboard-subscriptions-slack-performance/src/metabase/integrations/slack.clj#L132)

escherize committed:
Addressing code review issues

- Adds scheduled updates to the slack users and channels cache
  every 4 hours
- Removes `form_input_refresh` route
- Automatically refreshes cache when `form_input` route is hit,
  if 10 minutes have passed since the last cache fix
- Removes cache refresh on app startup
- Stores only user names, and channel names in the cache, not maps with
  irrelevant data
- Clears out some noisy linter warnings via `#_:clj-kondo/ignore`

escherize committed:
Adds slack task RefreshCache jobs

- Sets up startup-triggered and cron-triggered (every 4 hours at a
  random minute) slack channel/username cache refreshing

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into sharing-dashboard-subscriptions-slack-performance

escherize committed:
Implements 2 jobs to update slack users/channels

- one on startup, one every 4 hours
- improve docstrings

escherize commented on src/metabase/integrations/slack.clj:
```diff
@@ -135,55 +145,81 @@
-(defn channel-with-name
-  "Return a Slack channel with `channel-name` (as a map) if it exists."
+(defn channel-exists?
+  "Returns true if the channel it exists."
```
This function was only being used for this purpose before, so I made that expilcit here.

escherize committed:
ns linter fixes

escherize committed:
more ns linter fixes

escherize committed:
refactor files-channel to look in the cache

- instead of making a tricky api call
- fixing tests

escherize committed:
update files-channel-test to test the cache

escherize committed:
fixup integrations.slack ns decl

escherize committed:
fixup test integrations.slack ns decl

escherize committed:
fix outdated mock of slack/files-channel

escherize committed:
fix a few more tests

escherize committed:
fix api slack-test

escherize commented:
Thanks for getting your eyes on this PR, @noah and @cam. I've made some changes as requested, (and found some that weren't!). 

The new additions are:

- rename `slack-cache` to `slack-channel-and-user-cache setting`
- immediately strip out channel/user names from the slack response, and store those in the cache (rather than giant maps)
- integrate with the quartz scheduler to warm the cache on startup, and every 4 hours (which only does something when slack is configured)
- test the caching behavior in the slack api and integrations tests
- rework the `form_input` endpoint to refresh the cache in the background if it's been 10 minutes since an update (tracked via `slack-channels-and-usernames-last-updated` setting, so it will work with multiple instances)

camsaul commented on src/metabase/integrations/slack.clj:
```diff
@@ -135,84 +144,96 @@
+                           query-parameters)
        ;; filter out deleted users and bots. At the time of this writing there's no way to do this in the Slack API
        ;; itself so we need to do it after the fact.
        (filter (complement :deleted))
```
I wouldn't bother changing this now but for next time you can use `remove` e.g. `(remove :deleted)`. `remove` is basically the same as `filter` + `complement`

camsaul replied:
Oh nvm I'm dumb this isn't even part of your PR. Forget I said anything

escherize replied:
I fixed it in a subsequent commit anyway

camsaul commented on src/metabase/integrations/slack.clj:
```diff
@@ -135,84 +144,96 @@
+  "Refreshes users and conversations in slack-cache. If called with no args, finds both in parallel, sets
+  [[slack-cached-channels-and-usernames]], and resets the [[slack-channels-and-usernames-last-updated]] time."
+  []
+  (when (slack-configured?)
+    (log/info "Refreshing slack channels and usernames.")
+    (let [users (future (vec (users-list)))
+          conversations (future (vec (conversations-list)))]
+      (slack-cached-channels-and-usernames (concat @conversations @users))
+      (slack-channels-and-usernames-last-updated (t/zoned-date-time)))))
```
It might be worth considering putting some locking of some sort in here so we only attempt one update at a time for a given instance. If I make 100 calls to `refresh-channels-and-usernames!` and the first one triggers a refresh it would be nice if the other 99 just waited for the results of the first rather than issuing 100 Slack API requests. I would do something like this:

```clj
(defonce ^:private refresh-lock (Object.))

(defn- refresh-channels-and-usernames!* [] 
  ;; actual refresh logic goes here
  ...
  ;; update the last refresh time
  (slack-channels-and-usernames-last-updated (t/zoned-date-time)))

(defn- needs-refresh? []
  (< (slack-channels-and-usernames-last-updated) 10-minutes-ago))

(defn refresh-channels-and-usernames! []
  (when (needs-refresh?)
    (locking refresh-lock
      (when (needs-refresh?)
        (refresh-channels-and-usernames!*)))))
```

Basic idea is every time `refresh-channels-and-usernames!` is called, first double-check that a refresh is needed, by looking at the value of the `slack-channels-and-usernames-last-updated` Setting. If we actually need a refresh, then wait to acquire the `refresh-lock`; once we acquire that, check `slack-channels-and-usernames-last-updated` again in case it has been updated while we were waiting to get the lock. If we still need to refresh then we can do the actual refresh stuff, and update `slack-channels-and-usernames-last-updated`. Otherwise we can no-op.

Does that make sense?

camsaul replied:
I think that should be sufficient -- if we were really sophisticated we could try to avoid doing simultaneous refreshes across multiple instances but I think that's way more trouble than it's worth. As long as we limit ourselves to one refresh at a time per instance I think it's fine -- it will keep us from blowing up instances with tons of requests

camsaul reviewed (approved):
Overall this LGTM, I do think you should put a little bit of locking around the refresh logic so we can avoid multiple simultaneous refreshes on a single instance like I mentioned in my comment. But that's pretty straightforward so I would say go ahead and add that and then hit merge -- you don't need to hit me up for another review

escherize committed:
Merge branch 'master' into sharing-dashboard-subscriptions-slack-performance

escherize committed:
adds locking to refresh-channels-and-usernames!

escherize committed:
re-order requires
